### PR TITLE
Convert parts of core JS to typescript

### PR DESCRIPTION
**Pull request by charrondev**
As discussed this PR converts parts of core JS to typescript.

- JS and TS in compiled code works correctly.
- JS and TS in tested code works correctly.
- TSX works

The main utility functions have been converted to typescript here. This mostly consisted fixing some param descriptions and moving them inline. Not a lot of declarations needed annotations as it could all be inferred.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -19,6 +20,7 @@
         "setasap": "^2.0.0",
         "sprintf-js": "^1.1.1",
         "twemoji": "^2.5.1",
+        "typescript": "^2.8.1",
```
Comment (tburry) [inline package.json]: Is this not a devDependency?

File: src/scripts/__tests__/application.test.js [hunk 2]
```diff
@@ -4,38 +4,37 @@
...
-import * as application from "@core/application.js";
+import * as application from "@core/application";
 import gdn from "@core/gdn";
-import * as utility from "@core/utility";
 
 describe("metaDataFunctions", () => {
     it("return a default value if the requested one can't be found", () => {
         expect(application.getMeta("baz", "fallback")).toBe("fallback");
-    })
+    });
```
Comment (tburry) [inline src/scripts/__tests__/application.test.js]: Hallelujah

**Comment by tburry** [comment]
Haha you're a mad man.

**Comment by charrondev** [comment]
I've added a tslint-config [vanilla/tslint-config](https://github.com/vanilla/tslint-config). It has some similar catches to our eslint config but can actually be a good bit shorter because typescript is stricter by default.

This linting runs in travis as well.

**Pull request merged**